go_modules/lib/dependabot/go_modules/file_updater/go_mod_updater.rb

```diff
@@ -179,7 +161,12 @@ def run_go_get
 
           File.write(tmp_go_file, "package dummypkg\n") unless package
 
-          _, stderr, status = Open3.capture3(ENVIRONMENT, "go get -d")
+          # Use version pinning rather than `latest` just in case
+          # a new version gets released in the middle of our run.
+          version = "v" + dep.version.sub(/^v/i, "")
+          # TODO: go 1.18 will make `-d` the default behavior, so remove the flag then
+          command = "go get -d #{dep.name}@#{version}"
```

Comment:
I'm pondering the behaviour of this when there's a replacement applied to the module, e.g.
```
go 1.15

require github.com/pkg/errors v0.8.0
replace github.com/pkg/errors => github.com/thepwagner/errors v0.8.0
```


--- 

In upgrading to `0.8.1` (a version available in both repos), I'd expect Dependabot to output:
```
require github.com/pkg/errors v0.8.0
replace github.com/pkg/errors => github.com/thepwagner/errors v0.8.1
```



But `go get -d github.com/pkg/errors@v0.8.1` produces:
```
require github.com/pkg/errors v0.8.1
replace github.com/pkg/errors => github.com/thepwagner/errors v0.8.0
```

That has no effect, since the `github.com/thepwagner/errors v0.8.0` replacement is still applied.


And `go get -d github.com/thepwagner/errors@v0.8.1` produces:
```
require (
	github.com/pkg/errors v0.8.0
	github.com/thepwagner/errors v0.8.1 // indirect
)

replace github.com/pkg/errors => github.com/thepwagner/errors v0.8.0
```

That has no effect, since my code doesn't use that package.

---

FWIW, `updateDependencyFile` doesn't handle this correctly today either 😿 . I think https://github.com/thepwagner/action-update-go/blob/2ba2b9a3d49a3285b327622fbd5daccb4fe64e19/gomodules/applyupdate.go#L101-L136 is an implementation that does - and it looks closer to `updateDependencyFile` than `go get -d`.

Have you looked at `go mod edit` as an alternative to `updateDependencyFile`? It's not quite plug and play, but at least it's provided+supported by the SDK.

Reply:
> In upgrading to 0.8.1 (a version available in both repos), I'd expect Dependabot to output:
>```
> require github.com/pkg/errors v0.8.0
> replace github.com/pkg/errors => github.com/thepwagner/errors v0.8.1
>```

I actually disagree. My understanding of the _intent_ of the `replace` directive is that it's basically a "user override pin" that shouldn't be changed by automated tooling. I realize that some repos use it otherwise, but it's often either a [programmer mistake](https://github.com/dependabot/dependabot-core/pull/3576) or used to workaround specific problems with automated tooling... so if your automated tooling starts munging your workaround, then pretty soon you'll need a workaround for your workaround. 😀 

So in fact, your examples are exactly how I'd want and expect (based on the go docs/comments in the issue tracker) `go get` (and by extension Dependabot) to work. 

In fact, my mental model of Dependabot is that it merely automates the work of manually running `go get -u ./... && go mod tidy`. I get that it has a few additional tweaks such as ignore certain version updates, but in general the more we diverge from the typical `go` module update patterns, the more user confusion it'll probably create...

> Have you looked at go mod edit as an alternative to updateDependencyFile? It's not quite plug and play, but at least it's provided+supported by the SDK.

I haven't. And I doubt it will solve the root problem this PR is trying to solve where `go get` sees a mismatch between `go.mod` / `go.sum` and walks the dep tree and then blows up on the missing grpc module. Some of that may no longer be applicable when go lands lazy module loading, but very unclear at this point. 

Again, I say all this with less than 100% confidence in which outcomes of `go` machinery are expected vs bugs/inconsistencies that haven't been ironed out yet, but it's been very painful to debug this... The problematic repo that I am using for testing results in `go get -d` taking 20+ minutes for its initial cache population. So even to get this far has already taken more than a full day to narrow down what was happening in a semi-reproducible way.

Note also that I did see a big speed improvement when doing `go get -d <specific dep>` vs `go get -d`... I suspect the latter downloads all the modules specified in `go.mod`, vs the former only grabs a subset. So continuing with `go get -d` has some downsides.

Reply:
My use case here centres around long-lived forks:

As a worked example, let's say I want to add a feature to GitHub's API. To do this, I'm going to make a private fork of  https://github.com/google/go-github , and add my new functionality.
I don't want to leak the new feature into `go-github` before it's announced, so in the interim I'm going to use a `replace` directory to prefer my fork as the source of this dependency.

With my project in this setup, what happens if `google/go-github` makes a release?

Today, Dependabot is going to see the release of `google/go-github` and propose a PR performing that upgrade. Since it doesn't understand the `replace` directive, the resulting PR will be syntactically valid but semantically meaningless: merge or close, my project will still be using `thepwagner/go-github` version specified in the replace directive.

I don't think this PR will affect that behaviour.

---

What I'd prefer [in the scenario of a non-local replacement](https://github.com/thepwagner/action-update-go/tree/main/gomodules/testdata/replace) is:
1. Dependabot should consider the replaced dependency, (e.g. [`github.com/thepwagner/go-github`](https://github.com/thepwagner/action-update-go/blob/2ba2b9a3d49a3285b327622fbd5daccb4fe64e19/gomodules/dependencies_test.go#L33))
2. Dependabot should continue to propose updates, if available, from the fork. Ignoring the upstream.
3. Dependabot should update the [`replace` directive](https://github.com/thepwagner/action-update-go/blob/2ba2b9a3d49a3285b327622fbd5daccb4fe64e19/gomodules/applyupdate_test.go#L138-L155) and leave the `require` directives untouched (since they have no impact).

This isn't what `go get -u` does, but as a maintainer of a long-lived for it's what I'd like.

While this PR doesn't affect this scenario, switching from hand-updating the file to `go get -d` means (3) is potentially further than it was before. That was the flag I wanted to raise.

> go get -d <dep>@<version> is a clearly recommended way to bump to a particular version, so why not just use it if it solves the problems?

Agreed, SGTM!